File: backend/src/services/questionService.ts
import { QuestionGenerationResult, EvaluationResult } from '../types/index.js';
import { SYSTEM_PROMPT, buildQuestionPrompt, buildFollowupPrompt } from '../prompts/prompts.js';
import { getLLMProvider } from './llm/factory.js';
import { retrievalService } from './retrievalService.js';
import { z } from 'zod';

const QuestionResultSchema = z.object({
  reply: z.string(),
  day: z.number(),
  topic: z.string(),
  objective: z.string(),
  difficulty: z.string(),
  phase: z.string(),
});

export class QuestionService {
  async generateNextQuestion(params: {
    candidateName: string;
    jobRole: string;
    yearsExperience: number;
    currentDay: number;
    currentTopic: string;
    difficulty: string;
    phase: string;
    questionNumber: number;
    previousQuestions: string[];
    isFollowUp?: boolean;
    previousQuestion?: string;
    candidateAnswer?: string;
    lastEvaluation?: EvaluationResult;
  }): Promise<QuestionGenerationResult> {
    const llm = getLLMProvider();
    const dayInfo = retrievalService.getDayInfo(params.currentDay);
    const objectives = dayInfo?.objectives || [`Mastery of ${params.currentTopic}`];

    if (params.isFollowUp && params.previousQuestion && params.candidateAnswer && params.lastEvaluation) {
      const prompt = buildFollowupPrompt({
        candidateName: params.candidateName,
        previousQuestion: params.previousQuestion,
        candidateAnswer: params.candidateAnswer,
        tier: params.lastEvaluation.tier,
        missingConcepts: params.lastEvaluation.missingConcepts,
        currentDay: params.currentDay,
        currentTopic: params.currentTopic,
        difficulty: params.difficulty,
        phase: params.phase,
        previousQuestions: params.previousQuestions,
      });

      try {
        return await llm.generateStructured(prompt, SYSTEM_PROMPT, QuestionResultSchema);
      } catch (err) {
        console.warn('QuestionService followup LLM error, falling back:', err);
      }
    }

    const prompt = buildQuestionPrompt({
      candidateName: params.candidateName,
      jobRole: params.jobRole,
      yearsExperience: params.yearsExperience,
      currentDay: params.currentDay,
      currentTopic: params.currentTopic,
      objectives,
      difficulty: params.difficulty,
      phase: params.phase,
      questionNumber: params.questionNumber,
      previousQuestions: params.previousQuestions,
    });

    try {
      return await llm.generateStructured(prompt, SYSTEM_PROMPT, QuestionResultSchema);
    } catch (err) {
      console.warn('QuestionService question LLM error, falling back:', err);
      return {
        reply: `Regarding ${params.currentTopic}: how would you structure the architecture and handle key trade-offs in a production environment?`,
        day: params.currentDay,
        topic: params.currentTopic,
        objective: objectives[0] || `Mastery of Day ${params.currentDay}`,
        difficulty: params.difficulty,
        phase: params.phase,
      };
    }
  }
}

export const questionService = new QuestionService();

File: backend/app/services/question_service.py
from typing import List, Optional, Dict, Any
from app.schemas.schemas import QuestionGenerationResult, EvaluationResult
from app.prompts.prompts import SYSTEM_PROMPT, build_question_prompt, build_followup_prompt
from app.services.llm.factory import get_llm_provider
from app.services.retrieval_service import retrieval_service


class QuestionService:

    async def generate_next_question(
        self,
        candidate_name: str,
        job_role: str,
        years_experience: int,
        current_day: int,
        current_topic: str,
        difficulty: str,
        phase: str,
        question_number: int,
        previous_questions: List[str],
        is_follow_up: bool = False,
        previous_question: Optional[str] = None,
        candidate_answer: Optional[str] = None,
        last_evaluation: Optional[EvaluationResult] = None,
    ) -> QuestionGenerationResult:
        llm = get_llm_provider()
        day_info = retrieval_service.get_day_info(current_day)
        objectives = day_info.objectives if day_info else [f"Mastery of {current_topic}"]

        if is_follow_up and previous_question and candidate_answer and last_evaluation:
            prompt = build_followup_prompt(
                candidate_name=candidate_name,
                previous_question=previous_question,
                candidate_answer=candidate_answer,
                tier=last_evaluation.tier,
                missing_concepts=last_evaluation.missingConcepts,
                current_day=current_day,
                current_topic=current_topic,
                difficulty=difficulty,
                phase=phase,
                previous_questions=previous_questions,
            )
            try:
                return await llm.generate_structured(prompt, SYSTEM_PROMPT, QuestionGenerationResult)
            except Exception as err:
                print(f"QuestionService followup LLM error, falling back: {err}")

        prompt = build_question_prompt(
            candidate_name=candidate_name,
            job_role=job_role,
            years_experience=years_experience,
            current_day=current_day,
            current_topic=current_topic,
            objectives=objectives,
            difficulty=difficulty,
            phase=phase,
            question_number=question_number,
            previous_questions=previous_questions,
        )

        try:
            return await llm.generate_structured(prompt, SYSTEM_PROMPT, QuestionGenerationResult)
        except Exception as err:
            print(f"QuestionService question LLM error, falling back: {err}")
            return QuestionGenerationResult(
                reply=f"Regarding {current_topic}: how would you structure the architecture and handle key trade-offs in a production environment?",
                day=current_day,
                topic=current_topic,
                objective=objectives[0] if objectives else f"Mastery of Day {current_day}",
                difficulty=difficulty,
                phase=phase,
            )


question_service = QuestionService()
